omhttp: updates and improvements

gintrator opened the pull request:
<!--
LEGAL GDPR NOTICE:
According to the European data protection laws (GDPR), we would like to make you
aware that contributing to rsyslog via git will permanently store the
name and email address you provide as well as the actual commit and the
time and date you made it inside git's version history. This is inevitable,
because it is a main feature git. If you are concerned about your
privacy, we strongly recommend to use

--author "anonymous <gdpr@example.com>"

together with your commit. Also please do NOT sign your commit in this case,
as that potentially could lead back to you. Please note that if you use your
real identity, the GDPR grants you the right to have this information removed
later. However, we have valid reasons why we cannot remove that information
later on. The reasons are:

* this would break git history and make future merges unworkable
* the rsyslog projects has legitimate interest to keep a permanent record of the
  contributor identity, once given, for
  - copyright verification
  - being able to provide proof should a malicious commit be made

Please also note that your commit is public and as such will potentially be
processed by many third-parties. Git's distributed nature makes it impossible
to track where exactly your commit, and thus your personal data, will be stored
and be processed. If you would not like to accept this risk, please do either
commit anonymously or refrain from contributing to the rsyslog project.
-->
These changes aim to make omhttp a more general output module. Due to its origins as omelasticsearch, omhttp currently has a lot of elasticsearch-specific code to handle errors and batching. This has been changed to work more a wider range of services -- for example, error checking now just checks HTTP status codes rather than trying to parse an elasticsearch response JSON object.

Currently, omhttp only supports batching messages by concatenating with a newline character (this is the elasticsearch bulk format). I've implemented several batching strategies designed to work better with a wider array of REST-based services. Newline remains unchanged. Jsonarray builds an array of JSON messages, and Kafkarest allows the module to send to an HTTP Kafka REST endpoint.

Other implemented features include gzip compression, statistic counters for number of messages/requests/statuses, and retries via a retry ruleset when running in batch mode similar to omelasticsearch.

Documentation for updated functionality (including some design rationale w.r.t. retries in batching mode, more details on batching formats, example conf, etc..) in this PR: https://github.com/rsyslog/rsyslog-doc/pull/772.

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 58cb1ae9b00c60a79f1f08d2cc82962fcd03afc5 into 10b8235650d14285a7fcb658067e38807e9c45e9 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-0d34d6d596808b86b846a32bf4295f9d75ae4c02)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

rgerhards commented:
Thx for the PR. It seems to have some issues CI detected. Would you like to look into that or want a helping hand during this process? Just let me know.

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging b50a62d1e1fe29595d39148ab2d4a14295f738da into 040051a0087dbf8428c3ccc62f78d0a49daf0d83 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-d93348de32513b802b2cdfd17e1eef3b157dcd5a)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

gintrator commented:
I've fixed the Codefactor bash issues, but it looks like a new issue cropped up in the merge check with this error.
```
The following packages have unmet dependencies:
 libczmq-dev : Depends: libcurl4-nss-dev (>= 7.28.0) but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
```
This did not happen in the previous commit build, did something change?

gintrator commented:
Ah, looks like you are removing libczmq-dev in PR https://github.com/rsyslog/rsyslog/pull/3426, and my branch still has this code. Should I wait until that is merged and then merge those changes in here?

rgerhards commented:
> Ah, looks like you are removing libczmq-dev in PR #3426, and my branch still has this code. Should I wait until that is merged and then merge those changes in here?

sry for the late reply - it has been night over here ;-) Yes, please until #3426 is merged. As you have probably seen, it too seems to have detected some nits, which for some reason only surface after the removal of the 0mq libs. I am right now working on that.

rgerhards commented:
@gintrator I have re-initiated the failed Travis tests. Travis should pick up the changes from now-merged #3426. Please note that I may need to change the buildbot containers to run the new omhttp tests (which are very much appreciated!).

Side-note: `omhttp` is a contributed and orphaned module. Would you be willing to take it up? No real commitment required, but I would every now and then make you aware of upcoming questions and PRs... Just let me know.

rgerhards commented on tests/omhttp-basic.sh:
```diff
@@ -0,0 +1,45 @@
+omhttp_start_server -p $port
+
+#  Starting actual testbench
+. ${srcdir:=.}/diag.sh init
```
This needs to be the first thing the testbench does. As far as I can see, this test should also fail because `omhttp_start_server` is unknown at the time it is called.

gintrator replied:
Moved this to be the first thing each omhttp test does.

rgerhards commented on tests/omhttp-basic.sh:
```diff
@@ -0,0 +1,45 @@
+#  Starting actual testbench
+. ${srcdir:=.}/diag.sh init
+
+error_file=$(pwd)/rstb.omhttp.error.log
```
We cannot use fixed path, because test can (and do) run in parallel.

So this needs to be
```
error_file=$RSYSLOG_DYNNAME.omhttp.error.log
```

rgerhards commented on tests/omhttp-batch-jsonarray-fail-often.sh:
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash
+# This file is part of the rsyslog project, released under ASL 2.0
+port=8080
```
We need to decide if these tests here shall be able to run in parallel or not. Depends a bit on ressource usage. If running in parallel, we need dynamic port (there is plumbing for this), if not, we need to serialize via Makefile (like elasticsearch).

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 53b8c17942e455391dcf7090776d91876e102683 into 792f1e50e0bc2eac17cf7ee0b9a1a71c4a6b5707 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-d98adb82fead2f7ad417f66717c17ba18968674d)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 800f60442a8907cd2398dd607a494faa00354ba9 into 792f1e50e0bc2eac17cf7ee0b9a1a71c4a6b5707 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-37874c4227338ea2d2b44261fedf2dbc14e54e91)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 68645bc8401d1211445a210f0ffbb70ff9a276d8 into 792f1e50e0bc2eac17cf7ee0b9a1a71c4a6b5707 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-fbe71af2329e08245db791b3481ebc150eae6384)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

gintrator commented:
@rgerhards Working on getting the tests to run and pass. Made some changes to allow them to run in parallel.

I'd be happy to take up the `omhttp` module, at this point I am pretty familiar with its code base and I think it is a very useful module.

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging 71227d97bf9a34a01c3dddaab7158c4d2d3c168b into 792f1e50e0bc2eac17cf7ee0b9a1a71c4a6b5707 - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-ef46ee76178c7a69b7e176c79daf3a26ddcdbe3b)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

rgerhards commented:
please check the issues reported by LGTM, e.g. https://lgtm.com/projects/g/rsyslog/rsyslog/snapshot/bddedbc19bc9036b55cff38cc86fd8ad393ecd11/files/contrib/omhttp/omhttp.c?sort=name&dir=ASC&mode=heatmap#xe54fe2ca66aeb794:1

rgerhards commented:
> I'd be happy to take up the `omhttp` module, at this point I am pretty familiar with its code base and I think it is a very useful module.

Much appreciated and welcome! So let's see if we can get this into more widespread use. I'll also work with @jgerhards to see if we can get the module into the official package archives. Let's target everything for the 2019.03 release in March.

side-note: it looks like the Travis fail is "valid". I'll look into activating tests (if needed) later today or tomorrow (due to release building I may not make it today).

rgerhards commented:
@gintrator I have experimentally updated the Ubuntu18 CI container and manually initiated some re-builds. Let's see what we get...

rgerhards commented:
This pull request **introduces 1 alert** and **fixes 1** when merging c06c3b7da0ae81d669623152463aa203a37059db into ddacd3042131dbdec4abbfcb02a214f166b10e6a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-03bfad20bb75c962b65687eebf7132fb19d68db6)

**new alerts:**

* 1 for Comparison result is always the same

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

rgerhards commented:
This pull request **fixes 1 alert** when merging bd6e374e3772ff49319f1a8df152819432d31326 into ddacd3042131dbdec4abbfcb02a214f166b10e6a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-e2d980fb916d4035f53427e2a92c9529a3796765)

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

gintrator commented:
@rgerhards Cleaned up the LGTM alert and fixed the CI issues. The tests included with this PR run and pass. The `buildbot/rsyslog debian9 rsyslog` build failed on the last run, but it looks like it might just be a flaky result, would you mind taking a look?

rgerhards commented:
This pull request **fixes 1 alert** when merging ef869ec8d272f1071aabfc7e362bcf87984823f6 into ddacd3042131dbdec4abbfcb02a214f166b10e6a - [view on LGTM.com](https://lgtm.com/projects/g/rsyslog/rsyslog/rev/pr-cf855d09661582f126c9d97a23a275b68fdbdc1b)

**fixed alerts:**

* 1 for Commented\-out code

---

*Comment posted by [LGTM.com](https://lgtm.com)*

rgerhards commented:
There are indeed some flakes, have looked into them (and will continue to do so). Also see https://github.com/rsyslog/rsyslog/pull/3440

could you squash the commit to a useful set? Not sure if a single commit is the best thing to do.

gintrator commented:
Rebased the commits on top of master and squashed to 2 commits. One for changes to the module source and the other for testing related changes.

rgerhards commented:
@gintrator if possible, tests for this would be useful:

* retry ability, that is e.g. a) startup w/o http server running, b) starting it c) seeing it receives all data
* the compression capability
* authentication support?

I did check via CodeCov, which shows yet untested code pathes. There are some more areas that need test, but the three given are the biggest ones. If they were in, that would be excellent. Even without, the PR is still a veeery good step forward!

rgerhards commented on contrib/omhttp/omhttp.c:
```diff
@@ -144,27 +183,35 @@ static struct cnfparamdescr actpdescr[] = {
 	{ "dynrestpath", eCmdHdlrBinary, 0 },
-	{ "bulkmode", eCmdHdlrBinary, 0 },
+	{ "bulkmode", eCmdHdlrBinary, 0 }, /* batch and bulkmode are synonyms */
+	{ "batch", eCmdHdlrBinary, 0 },
```
why not stick with a single param? Two synonyms alwas cause trouble in the long term... Just a general thought.

gintrator replied:
Wanted to change terminology to be more general, since `bulkmode` is an elasticsearch-specific concept, while still being backward-compatible. Introduced the `batch.format`, `batch.maxbytes`, `batch.maxsize` here as well.

Agree that synonyms may be problematic long term. Would you prefer to keep the previous parameter names or drop them in favor of more nicely namespaced params?

rgerhards commented on contrib/omhttp/omhttp.c:
```diff
@@ -144,27 +183,35 @@ static struct cnfparamdescr actpdescr[] = {
+	{ "reloadonhup", eCmdHdlrBinary, 0 },
+	{ "retry", eCmdHdlrBinary, 0 }, /* retry and retryfailures are synonyms */
 	{ "retryfailures", eCmdHdlrBinary, 0 },
+	{ "retry.ruleset", eCmdHdlrString, 0 }, /* retry.ruleset and retryruleset are synonyms */
```
same here

rgerhards commented on contrib/omhttp/omhttp.c:
```diff
@@ -223,6 +293,9 @@ CODESTARTfreeInstance
 	free(pData->serverBaseUrls);
 	free(pData->uid);
+	free(pData->httpcontenttype);
+	if (pData->headerContentTypeBuf != NULL)
```
The `if`` is not required, `free(NULL)` is well-defined.

rgerhards commented on contrib/omhttp/omhttp.c:
```diff
@@ -233,7 +306,8 @@ CODESTARTfreeInstance
 	free(pData->checkPath);
 	free(pData->tplName);
-	free(pData->errorFile);
+	if (pData->errorFile != NULL)
```
same as above.

rgerhards commented on contrib/omhttp/omhttp.c:
```diff
@@ -452,7 +535,8 @@ checkConn(wrkrInstanceData_t *const pWrkrData)
 	if(urlBuf != NULL)
 		es_deleteStr(urlBuf);
-	free(pWrkrData->reply);
+	if (pWrkrData->reply != NULL)
```
NULL check not required

rgerhards commented on contrib/omhttp/omhttp.c:
```diff
@@ -530,786 +614,775 @@ setPostURL(wrkrInstanceData_t *const pWrkrData, uchar **const tpls)
 
 finalize_it:
-	free(cstr);
+	if (batchBuf != NULL)
```
NULL check not required

rgerhards commented on tests/Makefile.am:
```diff
@@ -718,6 +718,17 @@ TESTS +=  \
 endif
 
+if ENABLE_OMHTTP
+TESTS +=  \
```
it would be good to have test with valgrind support as well. They are easy to do, all three-liners:
```
#!/bin/bash
export USE_VALGRIND="YES"
source ${srcdir:=.}/non-valgrind-test.sh
```
Test files end in `-vg.sh`

gintrator replied:
Woops, overlooked this. I'll add a few valgrind tests.

rgerhards commented on tests/diag.sh:
```diff
@@ -1933,6 +1933,99 @@ init_elasticsearch() {
+
+    dep_work_dir=$(readlink -f .dep_wrk)
+    # Create work directory for parallel tests
+    omhttp_work_dir=${dep_work_dir}/omhttp_$RSYSLOG_DYNNAME
```
just general thought: `.dep*` is for external dependencies. These are NOT automatically cleaned up. It may be better to use `mkdir $RSYSLOG_DYNNAME/mydir` here - but it's up to discussion. The latter is cleaned up automatically at the end of successful test.

rgerhards commented on tests/omhttp-basic.sh:
```diff
@@ -0,0 +1,45 @@
+port="$(get_free_port)"
+omhttp_start_server $port
+
+error_file=$(pwd)/$RSYSLOG_DYNNAME.omhttp.error.log
```
```suggestion
error_file=$RSYSLOG_DYNNAME.omhttp.error.log
```
Leave the pwd part out - $RSYSLOG_DYNNAME is your sole prefix. Else things may break (now in some envs or later after a plumbing change).

rgerhards commented on tests/omhttp-basic.sh:
```diff
@@ -0,0 +1,45 @@
+
+generate_conf
+add_conf '
+#$DebugLevel 2
```
we don't need the debug level here. You can turn on debugging via `diag.sh`, see https://github.com/rsyslog/rsyslog/blob/master/tests/diag.sh#L75

rgerhards commented on tests/omhttp-basic.sh:
```diff
@@ -0,0 +1,45 @@
+if $msg contains "msgnum:" then
+	action(
+		# Payload
+		name="my_http_action"
```
just my personal preferrence: I put multiple parameters on one line as it makes it easier to grasp the test in one slurp - but **not** a requirement.

rgerhards commented on tests/omhttp-basic.sh:
```diff
@@ -0,0 +1,45 @@
+    )
+'
+startup
+injectmsg  0 10000
```
if you use
```
export NUMMESSAGES=10000
```
after test init, you do not need to specify the count in ìnjectmsg`and `seq_check` - makes it a bit easier to change the count (for whatever reason).

davidelang commented:
This is a new module (even if it is derived from an existing one), so there 
should be no need to have backwards compatibility.

This is notreplacing omelasticsearch, and if we choose to change omelasticsearch 
to be a wrapper around this module, that wrapper can do any backwards compatible 
variable manipulation.

so I think we should just drop the old names.

David Lang

rgerhards commented:
@gintrator FYI: I am enabling the compile of omhttp in more CI environments. I'll let you know if that brings something up.

rgerhards commented:
@gintrator @davidelang The module is **not** brand-new, BUT I think (almost) nobody so far used it. As the issue trackers show, it also seemd - in pre-this-PR state - not suitable for real use. So no problem from my side to remove old config params. I am not even sure if any doc for the module exists.

rgerhards commented:
@gintrator FYI: the CI failures look valid

gintrator commented:
@rgerhards I addressed some of your comments above -- added tests for the compression and basic auth capabilities, cleaned up the tests, cleaned up some issues in the C source that CI brought up, etc..

There are several retry tests already -- the omhttp test server is configured to throw 500 errors every n'th request, and we check to make sure all messages still make it through. I renamed these from "fail-often" to "retry". The 3 batch formats are tested, fixed up some issues Clang found in one of the newline tests. The latest CI run has a couple failures, but they look unrelated to this change?

Generally, I think we're at a pretty decent place now with these changes, do you have anything else you'd like to address?

gintrator commented:
@rgerhards One more thing I'd like to bring up. This module relies heavily on the output module [transaction interface](https://www.rsyslog.com/doc/v8-stable/development/dev_oplugins.html#batching-of-messages) to implement its internal batching feature (omhttp batch size <= rsyslog transaction size). Currently the transaction return code `RS_RET_DEFER_COMMIT` doesn't appear to work, as documented in Issue https://github.com/rsyslog/rsyslog/issues/2420. The `retry` and `retry.ruleset` parameters are necessary to ensure that if a batch of 50 messages fails, all 50 are still retried, instead of just the last message submitted. 

If this issue can be fixed, then this output module can play very nicely with the usual `action.resumeRetryCount` parameter while in batching mode (`batch="on"`), instead of relying on custom retry logic.

I'll keep an eye out for that, and please keep me in the loop if this is implemented, as I'd like to use that feature in omhttp at some point.

rgerhards commented:
> @rgerhards One more thing I'd like to bring up. This module relies heavily on the output module [transaction interface](https://www.rsyslog.com/doc/v8-stable/development/dev_oplugins.html#batching-of-messages) to implement its internal batching feature (omhttp batch size <= rsyslog transaction size). Currently the transaction return code `RS_RET_DEFER_COMMIT` doesn't appear to work, as documented in Issue #2420. The `retry` and `retry.ruleset` parameters are necessary to ensure that if a batch of 50 messages fails, all 50 are still retried, instead of just the last message submitted.

thx for adding this to the issue tracker. I'll try to address it some time not too late, but unfortunately right now is a very busy time. I'll do a review of the PR today.

darakian commented:
@rgerhards How long until we can expect this in master?

rgerhards commented:
I got hit by a bad flu after my last response. Re-visiting the tracker now, everything looks ready for merge, so this is what I will do. @gintrator thanks for your help and persistence.

mbonsack commented:
Is there an option to skip peer certificate validation when using https?  Like "curl -k"?
